### PR TITLE
fix(rlhf): mask teacher logprobs before KL

### PR DESCRIPTION
**Pull request by tongchen126**
# PR type

- [x] Bug Fix
- [ ] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

Fix NaN values in the OPD-RL `teacher_kl` monitoring metric when using padded batches.

Teacher API log probabilities use `0` outside completion tokens, while padding-free student log probabilities may use
`-1e10`. The K3 estimator previously applied the completion mask after exponentiation:

```python
d = teacher_logps - student_logps
k3 = (torch.exp(d) - d - 1) * completion_mask
```

At padding positions, this could produce:

```text
d = 0 - (-1e10) = 1e10
exp(d) = inf
inf * 0 = NaN
```

A single padding NaN then caused the aggregated `teacher_kl` metric to become NaN, even though all active
completion-token log probabilities were finite.

This PR masks the log-probability difference before applying the K3 estimator. The same safe masking is applied to the
signed teacher log-ratio so non-finite padding cannot contaminate OPD advantages.

The fix is implemented in the shared RL advantage helper, so it applies consistently to both Megatron and HF RLHF
trainers without duplicating logic in their training loops. Active completion-token calculations are unchanged.

Regression tests cover:

- The `-1e10` padding case that caused `inf * 0`.
- `NaN` and `Inf` values outside the completion mask.
- Preservation of the expected K3 and signed log-ratio values on active tokens.

## Experiment results

A minimal reproduction produced:

```text
delta = [[1.0, 1e10, -2.0, 1e10]]
k3 = [[0.7182817, NaN, 1.1353352, NaN]]
sum = NaN
```

After the fix, both regression tests pass:

```text
test_teacher_helpers_ignore_nonfinite_padding ... ok
test_teacher_kl_masks_padding_before_exp ... ok

Ran 2 tests
OK
```

Additional verification:

```text
Python compilation: passed
git diff --check: passed
```

### CODE REVIEW

**Comment by tongchen126** [comment]
Megatron RL log before the fix:
{'loss': 1.55689251, 'kl': 0.00021919, 'reward': 0.0, 'reward_std': 0.0, 'frac_reward_zero_std': 1.0, 'teacher_kl': nan, 'clip_ratio/low_mean': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/region_mean': 0.0, 'completions/mean_length': 134.0, 'completions/max_length': 147.25, 'completions/min_length': 120.75, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_max': 0.0, 'grad_norm': 4.88298607, 'learning_rate': 1.5e-07, 'load_balancing_loss': 1.45908439, 'iteration': '1/1302', 'elapsed_time': '2m 13s', 'remaining_time': '2d 0h 2m 38s', 'memory(GiB)': 86.6, 'train_speed(s/it)': 132.942599}

Megatron RL log after the fix:
{'loss': 1.55903983, 'kl': 0.00021919, 'reward': 0.0, 'reward_std': 0.0, 'frac_reward_zero_std': 1.0, 'teacher_kl': 1.38129056, 'clip_ratio/low_mean': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/region_mean': 0.0, 'completions/mean_length': 134.0, 'completions/max_length': 147.25, 'completions/min_length': 120.75, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_max': 0.0, 'grad_norm': 4.91996193, 'learning_rate': 1.5e-07, 'load_balancing_loss': 1.45908439, 'iteration': '1/1302', 'elapsed_time': '2m 51s', 'remaining_time': '2d 13h 47m 21s', 'memory(GiB)': 86.6, 'train_speed(s/it)': 170.977078}

File: swift/rl_core/advantage.py [hunk 1]
```diff
@@ -222,8 +222,10 @@ def compute_teacher_kl_per_token(
         ``[B, T]`` per-token teacher KL (masked outside the response).
     """
     d = teacher_per_token_logps - policy_per_token_logps
+    # Mask before exp so padding sentinel values cannot overflow and produce inf * 0.
+    d = d.masked_fill(~completion_mask.bool(), 0.0)
```
[the remark this answers is not in the document]
Comment (tongchen126) [reply swift/rl_core/advantage.py:226]: it is not necessary in the current code path.

  completion_mask.bool()

  converts the mask to Boolean while preserving its device. Teacher logprobs, student logprobs, and completion_mask are already constructed on the same
  device.

File: swift/rl_core/advantage.py [hunk 2]
```diff
@@ -245,7 +247,7 @@ def compute_teacher_logratio(
         ``[B, T]`` per-token signed log-ratio (masked outside the response).
     """
     d = teacher_per_token_logps - policy_per_token_logps
-    return d * completion_mask
+    return d.masked_fill(~completion_mask.bool(), 0.0)
```
[the remark this answers is not in the document]
Comment (tongchen126) [reply swift/rl_core/advantage.py:250]: it is not necessary in the current code path.

  completion_mask.bool()

  converts the mask to Boolean while preserving its device. Teacher logprobs, student logprobs, and completion_mask are already constructed on the same
  device.

**Review by hjh0119 — APPROVED** [review]

**Comment by hjh0119** [comment]
thanks
